### ISSUE TITLE
Question about the Meaning of Comparing z_logits and z_t in c_loss_1 Calculation

### ISSUE THREAD

**Issue by inanb** [Yu-Fangxu/COLD-Attack#14]
Hello,

I have some doubts about c_loss_1 and would like to consult you.

In `xyz_logits, xy_length = soft_forward_xyz(model, soft_forward_x, soft_forward_y_, z_onehot)`, `xyz_logits` is obtained by splicing `soft_forward_x`, `soft_forward_y_`, and `z_onehot` as inputs to the model for one inference to get the output `xyz_logits`. And `z_logits = torch.cat([xyz_logits[bi * lg + st:bi * lg + ed, :] for bi in range(bz)], dim=0)` shows that `z_logits` comes from `xyz_logits`. c_loss_1 compares `z_logits` with `z_t`.

However, according to the above description, `z_logits` is the logits obtained by inference with x, y, and z as inputs. So what does it mean to compare `z_logits` and `z_t`?

**Comment by Yu-Fangxu** [comment]
Hi, 
In this context, z_logits refers to the logits calculated based on the context provided by x and y, rather than the content generated after xyz. The purpose of comparing z_logits with z_t (sure, here is ...) is to align z_logits, given the context of x and y, with the intended target of the attack (sure, here is ...).

**Comment by inanb** [comment]
Thank you for your prompt reply.

I noticed that xyz_embeds is obtained from embed_inputs with y_logits, x_onehot, and z_onehot passed in.
```python
def soft_forward_xyz(model, x_onehot, y_logits, z_onehot):
    '''
    computes logits for $y$, based on a fixed context $y$ and the current logit distribution of $y$
    :param model:
    :param x_onehot:
    :param y_logits:
    :return:
    '''
    xyz_embeds = embed_inputs(
        model.get_input_embeddings().weight,
        y_logits,
        x_onehot=x_onehot,
        z_onehot=z_onehot,
        device=y_logits.device
    )
    xyz_logits = model(inputs_embeds=xyz_embeds).logits
    if x_onehot is not None:
        xy_length = x_onehot.shape[1] + y_logits.shape[1]
    else:
        xy_length = y_logits.shape[1]
    return xyz_logits, xy_length
``` 

And below, I observed that the concatenation of probs should be prob = x_onehot + F.softmax(y_logits, dim=-1).type(torch.float16) + z_onehot. Finally, the embedding of prob is returned, which is xyz_logits. This makes me think that the input is the inference input of x + y + z, and the output is z_logits.

```python
def embed_inputs(embedding, logits, x_onehot=None, z_onehot=None, device='cuda'):
    '''
    embeds inputs in a dense representation, before passing them to the model
    '''
    # typically we embed a one-hot vector. But here since we work we work with dense representations,
    # we have softmax here to make sure that all the values of the input logits sum to one (similar to a 1-hot vector).
    probs = F.softmax(logits, dim=-1).type(torch.float16)
    # embedding : [vocab_size, embedding_size]
    # logits:     [batch_size, length, vocab_size]
    if x_onehot is not None:
        probs = torch.cat((x_onehot.type(torch.FloatTensor), probs.type(torch.FloatTensor)), dim=1)
    if z_onehot is not None:
        probs = torch.cat((probs.type(torch.FloatTensor), z_onehot.type(torch.FloatTensor)), dim=1)
        
    probs = probs.to(device)
    return torch.matmul(probs, embedding)
``` 

In fact, I think your statement is correct. I also thought so after reading the paper. 

However, maybe due to my insufficient coding skills, I might have a wrong understanding of the above functions. I hope you can help me clarify this. Thank you🙏

**Comment by inanb** [comment]
Thank you for your prompt reply.

I noticed that xyz_embeds is obtained from embed_inputs with y_logits, x_onehot, and z_onehot passed in.
```python
def soft_forward_xyz(model, x_onehot, y_logits, z_onehot):
    '''
    computes logits for $y$, based on a fixed context $y$ and the current logit distribution of $y$
    :param model:
    :param x_onehot:
    :param y_logits:
    :return:
    '''
    xyz_embeds = embed_inputs(
        model.get_input_embeddings().weight,
        y_logits,
        x_onehot=x_onehot,
        z_onehot=z_onehot,
        device=y_logits.device
    )
    xyz_logits = model(inputs_embeds=xyz_embeds).logits
    if x_onehot is not None:
        xy_length = x_onehot.shape[1] + y_logits.shape[1]
    else:
        xy_length = y_logits.shape[1]
    return xyz_logits, xy_length
``` 

And below, I observed that the concatenation of probs should be prob = x_onehot + F.softmax(y_logits, dim=-1).type(torch.float16) + z_onehot. Finally, the embedding of prob is returned, which is xyz_logits. This makes me think that the input is the inference input of x + y + z, and the output is z_logits.

```python
def embed_inputs(embedding, logits, x_onehot=None, z_onehot=None, device='cuda'):
    '''
    embeds inputs in a dense representation, before passing them to the model
    '''
    # typically we embed a one-hot vector. But here since we work we work with dense representations,
    # we have softmax here to make sure that all the values of the input logits sum to one (similar to a 1-hot vector).
    probs = F.softmax(logits, dim=-1).type(torch.float16)
    # embedding : [vocab_size, embedding_size]
    # logits:     [batch_size, length, vocab_size]
    if x_onehot is not None:
        probs = torch.cat((x_onehot.type(torch.FloatTensor), probs.type(torch.FloatTensor)), dim=1)
    if z_onehot is not None:
        probs = torch.cat((probs.type(torch.FloatTensor), z_onehot.type(torch.FloatTensor)), dim=1)
        
    probs = probs.to(device)
    return torch.matmul(probs, embedding)
``` 

In fact, I think your statement is correct. I also thought so after reading the paper. However, maybe due to my insufficient coding skills, I might have a wrong understanding of the above functions. I hope you can help me clarify this. Thank you!

**Comment by Yu-Fangxu** [comment]
Hi,
Here embed_inputs() returns token embeddings rather than token logits. ```F.softmax(y_logits, dim=-1)``` aims to approximate y as discrete tokens like x and z, then we use ```torch.matmul(probs, embedding)``` to approximate the embeddings of discrete tokens because it's hard to input the logits to the LLMs directly. Finally, the output from ```model(inputs_embeds=xyz_embeds).logits``` computes the logits of xyz. ```z_logits = torch.cat([xyz_logits[bi * lg + st:bi * lg + ed, :] for bi in range(bz)], dim=0)```, which looks a little bit complicated, it works to select logits correspond to z from xyz_logits in a batch . It would be great if my answer helps you, thanks!

**Comment by inanb** [comment]
Thank you so much! I think I kind of get it now.
Could you please kindly explain to me again why we need to concatenate z_onehot? If the purpose is to align with z, wouldn't it be enough to just fix x and y (x_onehot + F.softmax(y_logits, dim=-1))? Why do we have to concatenate z_onehot as well (x_onehot + F.softmax(y_logits, dim=-1).type(torch.float16) + z_onehot)?

**Comment by Yu-Fangxu** [comment]
Sure, this is because the output logits is regarding xyz, if you remove z_onehot, the output logits correspond to x + y. We expect the extension of x + y to be z, if we only have xy, it's hard to compute p(z|x, y) directly, we need to put z in the context, then we can know how large is the probability of LLM can predict z, then we enforce this probability to be as much as larger given y.

**Comment by inanb** [comment]
Perhaps I've figured out where our differences in understanding lie.
I think that in soft_forward_xyz, xyz_logits = model(inputs_embeds=xyz_embeds).logits means using xyz_embeds as the input to obtain the logits of the next sequence. However, you seem to think there's some other meaning to it.
Could you please kindly explain this to me? I really can't come up with any other implications besides predicting the next sequence or next token😭.

**Comment by inanb** [comment]
I think I've got it now.
It can be simply regarded as the score. After providing the xyz context, we can consider the logits as scores and then align them with the ideal output. Maybe this way of understanding is correct. Thank you so much for your answers and your work!

**Comment by vietvo89** [comment]
hi @inanb and @Yu-Fangxu . Did you see the loss is `nan`? I followed the instruction in `README.md`. I ran `bash attack.sh "suffix"`. I printed the `loss` for every iteration.

The outputs look like below.

```
1, loss: -911.9691,flu_loss: 0.5084, c_loss_1: 3.8594, c_loss_2: 11.8672, lr: 0.1000, |. It' </s></s>...   I'm just an AI, it'. |
iter=2, loss: nan,flu_loss: nan, c_loss_1: nan, c_loss_2: nan, lr: 0.1000
iter=3, loss: nan,flu_loss: nan, c_loss_1: nan, c_loss_2: nan, lr: 0.1000
iter=4, loss: nan,flu_loss: nan, c_loss_1: nan, c_loss_2: nan, lr: 0.1000
iter=5, loss: nan,flu_loss: nan, c_loss_1: nan, c_loss_2: nan, lr: 0.1000
iter=6, loss: nan,flu_loss: nan, c_loss_1: nan, c_loss_2: nan, lr: 0.1000
iter=7, loss: nan,flu_loss: nan, c_loss_1: nan, c_loss_2: nan, lr: 0.1000
iter=8, loss: nan,flu_loss: nan, c_loss_1: nan, c_loss_2: nan, lr: 0.1000
iter=9, loss: nan,flu_loss: nan, c_loss_1: nan, c_loss_2: nan, lr: 0.1000
```
